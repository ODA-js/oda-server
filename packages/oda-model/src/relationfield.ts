import { merge } from 'lodash';
import { Nullable, assignValue, MetaModelType } from './types';
import { RelationInput } from './relation';
import { HasOne } from './hasone';
import { HasMany } from './hasmany';
import { BelongsToMany } from './belongstomany';
import { BelongsTo } from './belongsto';
import {
  RelationFieldBase,
  RelationFieldBaseMetaInfo,
  RelationFieldBasePersistence,
  IRelationFieldBase,
  RelationFieldBaseInternal,
  RelationFieldBaseInput,
  RelationFieldBaseOutput,
} from './relationfieldbase';
import { Internal } from './element';

function discoverFieldType(obj: any) {
  // сделать проверку по полю...
  if (obj.hasOne) {
    return 'HasOne';
  } else if (obj.hasMany) {
    return 'HasMany';
  } else if (obj.belongsTo) {
    return 'BelongsTo';
  } else if (obj.belongsToMany) {
    return 'BelongsToMany';
  } else {
    console.warn(`undefined relation type of ${JSON.stringify(obj)}`);
    return 'undefined';
  }
}

export interface RelationFieldMetaInfo
  extends RelationFieldBaseMetaInfo<RelationFieldPersistence> {}

export interface RelationFieldPersistence
  extends RelationFieldBasePersistence {}

export interface RelationFieldInput
  extends RelationFieldBaseInput<
    RelationFieldMetaInfo,
    RelationFieldPersistence
  > {
  relation: RelationInput;
}

export interface RelationFieldOutput
  extends RelationFieldBaseOutput<
    RelationFieldMetaInfo,
    RelationFieldPersistence
  > {
  relation: RelationInput;
}

/**
 * relation field definition
 */
export interface IRelationField
  extends IRelationFieldBase<
    RelationFieldMetaInfo,
    RelationFieldInput,
    RelationFieldPersistence,
    RelationFieldOutput
  > {}

export interface RelationFieldInput
  extends RelationFieldBaseInput<
    RelationFieldMetaInfo,
    RelationFieldPersistence
  > {}

export interface RelationFieldInternal extends RelationFieldBaseInternal {}

const defaultMetaInfo = {};
const defaultInput = {
  metadata: defaultMetaInfo,
  persistent: false,
  derived: false,
  identity: false,
  indexed: false,
  required: false,
};

export class RelationField
  extends RelationFieldBase<
    RelationFieldMetaInfo,
    RelationFieldInput,
    RelationFieldInternal,
    RelationFieldPersistence,
    RelationFieldOutput
  >
  implements IRelationField {
  public get modelType(): MetaModelType {
    return 'relation-field';
  }
  constructor(init: RelationFieldInput) {
    super(merge({}, defaultInput, init));
  }

  public updateWith(input: Nullable<RelationFieldInput>) {
    super.updateWith(input);

    assignValue<
      RelationFieldInternal,
      RelationFieldInput,
      RelationFieldInput['relation']
    >({
      src: this[Internal],
      input,
      field: 'relation',
      effect: (src, value) => {
        switch (discoverFieldType(value)) {
          case 'HasOne':
            src.relation = new HasOne({
              ...(value as { hasOne: string }),
              entity: this.metadata.entity,
              field: this.name,
            });
            break;
          case 'HasMany':
            src.relation = new HasMany({
              ...(value as { hasMany: string }),
              entity: this.metadata.entity,
              field: this.name,
            });
            break;
          case 'BelongsToMany':
            src.relation = new BelongsToMany({
              ...(value as { belongsToMany: string; using: string }),
              entity: this.metadata.entity,
              field: this.name,
            });
            break;
          case 'BelongsTo':
            src.relation = new BelongsTo({
              ...(value as { belongsTo: string }),
              entity: this.metadata.entity,
              field: this.name,
            });
            break;
          default:
            throw new Error('undefined type');
        }
      },
    });

    assignValue<
      RelationFieldMetaInfo,
      RelationFieldInput,
      NonNullable<RelationFieldInput['derived']>
    >({
      src: this.metadata,
      input,
      inputField: 'derived',
      effect: (src, value) => {
        if (
          this.relation.verb === 'HasOne' ||
          this.relation.verb === 'HasMany'
        ) {
          src.persistence.derived = value;
        } else {
          src.persistence.derived = false;
        }
      },
      required: true,
      setDefault: src => (src.persistence.derived = defaultInput.derived),
    });

    assignValue<
      RelationFieldMetaInfo,
      RelationFieldInput,
      NonNullable<RelationFieldInput['persistent']>
    >({
      src: this.metadata,
      input,
      inputField: 'persistent',
      effect: (src, value) => {
        if (this.relation.verb === 'BelongsTo') {
          src.persistence.persistent = value;
        } else {
          src.persistence.persistent = false;
        }
      },
      required: true,
      setDefault: src => (src.persistence.persistent = defaultInput.persistent),
    });

    assignValue<
      RelationFieldMetaInfo,
      RelationFieldInput,
      NonNullable<RelationFieldInput['identity']>
    >({
      src: this.metadata,
      input,
      inputField: 'identity',
      effect: (src, value) => {
        if (this.relation.verb === 'BelongsTo') {
          src.persistence.identity = value;
        } else {
          src.persistence.identity = false;
        }
      },
      required: true,
      setDefault: src => (src.persistence.identity = defaultInput.identity),
    });

    assignValue<
      RelationFieldMetaInfo,
      RelationFieldInput,
      NonNullable<RelationFieldInput['indexed']>
    >({
      src: this.metadata,
      input,
      inputField: 'indexed',
      effect: (src, value) => {
        if (this.relation.verb === 'BelongsTo') {
          src.persistence.indexed = value;
        } else {
          src.persistence.indexed = false;
        }
      },
      required: true,
      setDefault: src => (src.persistence.indexed = defaultInput.indexed),
    });

    assignValue<
      RelationFieldMetaInfo,
      RelationFieldInput,
      NonNullable<RelationFieldInput['required']>
    >({
      src: this.metadata,
      input,
      inputField: 'required',
      effect: (src, value) => {
        if (this.relation.verb === 'BelongsTo') {
          src.persistence.required = value;
        } else {
          src.persistence.required = false;
        }
      },
      required: true,
      setDefault: src => (src.persistence.required = defaultInput.required),
    });
  }

  // it get fixed object
  public toObject(): RelationFieldOutput {
    return merge({}, super.toObject(), {
      relation: this.relation.toObject(),
    } as Partial<RelationFieldOutput>);
  }

  public mergeWith(payload: Nullable<RelationFieldInput>) {
    super.mergeWith(payload);
  }
}
