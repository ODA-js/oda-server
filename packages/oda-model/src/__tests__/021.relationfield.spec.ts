import 'jest';
import { RelationField } from '../relationfield';

describe('RelationField', () => {
  it('derived & persistent & required & indexed & identity', () => {
    const notRelated = new RelationField({
      name: 'PD',
      derived: true,
      persistent: true,
      identity: true,
      indexed: true,
      required: true,
      relation: {
        hasOne: 'D#',
      },
    });
    expect(notRelated.persistent).toBeFalsy();
    expect(notRelated.derived).toBeTruthy();
    expect(notRelated.identity).toBeFalsy();
    expect(notRelated.indexed).toBeFalsy();
    expect(notRelated.required).toBeFalsy();

    const related = new RelationField({
      name: 'PD',
      derived: true,
      persistent: true,
      identity: true,
      indexed: true,
      required: true,
      relation: {
        belongsTo: 'D#',
      },
    });
    expect(related.persistent).toBeTruthy();
    expect(related.derived).toBeFalsy();
    expect(related.identity).toBeTruthy();
    expect(related.indexed).toBeTruthy();
    expect(related.required).toBeTruthy();
  });
  it('hasMany', () => {
    const res = new RelationField({
      name: 'ref',
      relation: {
        hasMany: 'A#',
      },
    });
    expect(res).toMatchSnapshot('hasMany');
    expect(res.toObject()).toMatchSnapshot('hasMany to Object');
  });
  it('hasOne', () => {
    const res = new RelationField({
      name: 'ref',
      relation: {
        hasOne: 'A#',
      },
    });
    expect(res).toMatchSnapshot('hasOne');
    expect(res.toObject()).toMatchSnapshot('hasOne to Object');
  });
  it('belongsTo', () => {
    const res = new RelationField({
      name: 'ref',
      relation: {
        belongsTo: 'A#',
      },
    });
    expect(res).toMatchSnapshot('belongsTo');
    expect(res.toObject()).toMatchSnapshot('belongsTo to Object');
  });
  it('belongsToMany', () => {
    const res = new RelationField({
      name: 'ref',
      relation: {
        belongsToMany: 'A#',
        using: '_B#',
      },
    });
    expect(res).toMatchSnapshot('belongsToMany');
    expect(res.toObject()).toMatchSnapshot('belongsToMany to Object');
  });
});
