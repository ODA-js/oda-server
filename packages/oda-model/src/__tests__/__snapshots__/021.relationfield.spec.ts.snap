// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RelationField belongsTo: belongsTo 1`] = `
RelationField {
  Symbol(internal): Object {
    "args": Map {},
    "name": "ref",
    "relation": BelongsTo {
      Symbol(internal): Object {
        "belongsTo": EntityReference {
          Symbol(internal): Object {
            "entity": "A",
            "field": "id",
          },
          Symbol(metadata): Object {},
        },
        "field": "ref",
        "name": "BelongsToRef",
      },
      Symbol(metadata): Object {
        "name": Object {
          "full": "BelongsToRef",
          "normal": "Ref",
          "short": "Ref",
        },
        "persistence": Object {
          "embedded": false,
          "single": true,
          "stored": true,
        },
        "verb": "BelongsTo",
      },
    },
  },
  Symbol(metadata): Object {
    "acl": Object {
      "create": Array [],
      "delete": Array [],
      "read": Array [],
      "update": Array [],
    },
    "persistence": Object {
      "derived": false,
      "identity": false,
      "indexed": false,
      "persistent": true,
      "required": false,
    },
  },
}
`;

exports[`RelationField belongsTo: belongsTo to Object 1`] = `
Object {
  "args": Object {},
  "description": undefined,
  "inheritedFrom": undefined,
  "metadata": Object {
    "acl": Object {
      "create": Array [],
      "delete": Array [],
      "read": Array [],
      "update": Array [],
    },
    "persistence": Object {
      "derived": false,
      "identity": false,
      "indexed": false,
      "persistent": true,
      "required": false,
    },
  },
  "name": "ref",
  "relation": Object {
    "belongsTo": "A#id",
    "entity": undefined,
    "field": "ref",
    "metadata": Object {
      "name": Object {
        "full": "BelongsToRef",
        "normal": "Ref",
        "short": "Ref",
      },
      "persistence": Object {
        "embedded": false,
        "single": true,
        "stored": true,
      },
      "verb": "BelongsTo",
    },
    "name": "BelongsToRef",
    "opposite": undefined,
  },
  "title": undefined,
}
`;

exports[`RelationField belongsToMany: belongsToMany 1`] = `
RelationField {
  Symbol(internal): Object {
    "args": Map {},
    "name": "ref",
    "relation": BelongsToMany {
      Symbol(internal): Object {
        "belongsToMany": EntityReference {
          Symbol(internal): Object {
            "backField": "id",
            "entity": "A",
            "field": "id",
          },
          Symbol(metadata): Object {},
        },
        "field": "ref",
        "fields": Map {},
        "name": "BelongsToManyRefs",
        "using": EntityReference {
          Symbol(internal): Object {
            "backField": "id",
            "entity": "B",
            "field": "id",
          },
          Symbol(metadata): Object {},
        },
      },
      Symbol(metadata): Object {
        "name": Object {
          "full": "BelongsToManyRefs",
          "normal": "Refs",
          "short": "Refs",
        },
        "persistence": Object {
          "embedded": false,
          "single": false,
          "stored": false,
        },
        "verb": "BelongsToMany",
      },
    },
  },
  Symbol(metadata): Object {
    "acl": Object {
      "create": Array [],
      "delete": Array [],
      "read": Array [],
      "update": Array [],
    },
    "persistence": Object {
      "derived": false,
      "identity": false,
      "indexed": false,
      "persistent": false,
      "required": false,
    },
  },
}
`;

exports[`RelationField belongsToMany: belongsToMany to Object 1`] = `
Object {
  "args": Object {},
  "description": undefined,
  "inheritedFrom": undefined,
  "metadata": Object {
    "acl": Object {
      "create": Array [],
      "delete": Array [],
      "read": Array [],
      "update": Array [],
    },
    "persistence": Object {
      "derived": false,
      "identity": false,
      "indexed": false,
      "persistent": false,
      "required": false,
    },
  },
  "name": "ref",
  "relation": Object {
    "belongsToMany": "id@A#id",
    "entity": undefined,
    "field": "ref",
    "fields": Array [],
    "metadata": Object {
      "name": Object {
        "full": "BelongsToManyRefs",
        "normal": "Refs",
        "short": "Refs",
      },
      "persistence": Object {
        "embedded": false,
        "single": false,
        "stored": false,
      },
      "verb": "BelongsToMany",
    },
    "name": "BelongsToManyRefs",
    "opposite": undefined,
    "using": "id@B#id",
  },
  "title": undefined,
}
`;

exports[`RelationField hasMany: hasMany 1`] = `
RelationField {
  Symbol(internal): Object {
    "args": Map {},
    "name": "ref",
    "relation": HasMany {
      Symbol(internal): Object {
        "field": "ref",
        "hasMany": EntityReference {
          Symbol(internal): Object {
            "backField": "id",
            "entity": "A",
            "field": "id",
          },
          Symbol(metadata): Object {},
        },
        "name": "HasManyRefs",
      },
      Symbol(metadata): Object {
        "name": Object {
          "full": "HasManyRefs",
          "normal": "Refs",
          "short": "Refs",
        },
        "persistence": Object {
          "embedded": false,
          "single": false,
          "stored": false,
        },
        "verb": "HasMany",
      },
    },
  },
  Symbol(metadata): Object {
    "acl": Object {
      "create": Array [],
      "delete": Array [],
      "read": Array [],
      "update": Array [],
    },
    "persistence": Object {
      "derived": false,
      "identity": false,
      "indexed": false,
      "persistent": false,
      "required": false,
    },
  },
}
`;

exports[`RelationField hasMany: hasMany to Object 1`] = `
Object {
  "args": Object {},
  "description": undefined,
  "inheritedFrom": undefined,
  "metadata": Object {
    "acl": Object {
      "create": Array [],
      "delete": Array [],
      "read": Array [],
      "update": Array [],
    },
    "persistence": Object {
      "derived": false,
      "identity": false,
      "indexed": false,
      "persistent": false,
      "required": false,
    },
  },
  "name": "ref",
  "relation": Object {
    "entity": undefined,
    "field": "ref",
    "hasMany": "id@A#id",
    "metadata": Object {
      "name": Object {
        "full": "HasManyRefs",
        "normal": "Refs",
        "short": "Refs",
      },
      "persistence": Object {
        "embedded": false,
        "single": false,
        "stored": false,
      },
      "verb": "HasMany",
    },
    "name": "HasManyRefs",
    "opposite": undefined,
  },
  "title": undefined,
}
`;

exports[`RelationField hasOne: hasOne 1`] = `
RelationField {
  Symbol(internal): Object {
    "args": Map {},
    "name": "ref",
    "relation": HasOne {
      Symbol(internal): Object {
        "field": "ref",
        "hasOne": EntityReference {
          Symbol(internal): Object {
            "backField": "id",
            "entity": "A",
            "field": "id",
          },
          Symbol(metadata): Object {},
        },
        "name": "HasOneRef",
      },
      Symbol(metadata): Object {
        "name": Object {
          "full": "HasOneRef",
          "normal": "Ref",
          "short": "Ref",
        },
        "persistence": Object {
          "embedded": false,
          "single": true,
          "stored": false,
        },
        "verb": "HasOne",
      },
    },
  },
  Symbol(metadata): Object {
    "acl": Object {
      "create": Array [],
      "delete": Array [],
      "read": Array [],
      "update": Array [],
    },
    "persistence": Object {
      "derived": false,
      "identity": false,
      "indexed": false,
      "persistent": false,
      "required": false,
    },
  },
}
`;

exports[`RelationField hasOne: hasOne to Object 1`] = `
Object {
  "args": Object {},
  "description": undefined,
  "inheritedFrom": undefined,
  "metadata": Object {
    "acl": Object {
      "create": Array [],
      "delete": Array [],
      "read": Array [],
      "update": Array [],
    },
    "persistence": Object {
      "derived": false,
      "identity": false,
      "indexed": false,
      "persistent": false,
      "required": false,
    },
  },
  "name": "ref",
  "relation": Object {
    "entity": undefined,
    "field": "ref",
    "hasOne": "id@A#id",
    "metadata": Object {
      "name": Object {
        "full": "HasOneRef",
        "normal": "Ref",
        "short": "Ref",
      },
      "persistence": Object {
        "embedded": false,
        "single": true,
        "stored": false,
      },
      "verb": "HasOne",
    },
    "name": "HasOneRef",
    "opposite": undefined,
  },
  "title": undefined,
}
`;
